akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala

```diff
@@ -0,0 +1,291 @@
+/**
+  * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+  */
+
+package akka.http.impl.engine.client
+
+import akka.http.impl.engine.client.RedirectSupportStage.RedirectMapper
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.settings.ClientAutoRedirectSettings
+import akka.http.scaladsl.settings.ClientAutoRedirectSettings.HeadersForwardMode.{All, Only, Zero}
+import akka.stream._
+import akka.stream.scaladsl.BidiFlow
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+
+import scala.language.higherKinds
+import scala.util.control.NoStackTrace
+
+object RedirectSupportStage {
+  type RedirectMapper = (ClientAutoRedirectSettings, HttpRequest, HttpResponse) => Option[HttpRequest]
+
+  case object InfiniteRedirectLoopException
+    extends RuntimeException("Infinite redirect loop detected, breaking.")
+      with NoStackTrace
+
+  def inferMethod(req: HttpRequest, resp: HttpResponse): Option[HttpMethod] = {
+    import HttpMethods._
+    import StatusCodes._
+    val m = req.method
+    resp.status match {
+      case MovedPermanently | Found | TemporaryRedirect | PermanentRedirect if m == GET || m == HEAD => Some(m)
+      case SeeOther => Some(GET)
+      case _ => None
+    }
+  }
+
+  val defaultRedirectMapper: RedirectMapper =
+    (s, req, resp) => {
+      if (resp.status.isRedirection) {
+        inferMethod(req, resp).flatMap { method =>
+          resp.headers.find(_.is("location")).flatMap(location => {
+            val item = if (sameOrigin(req.uri, Uri(location.value))) s.sameOrigin else s.crossOrigin
+            if (item.getAllow) {
+              val headers = item.getForwardHeaders match {
+                case All => req.headers
+                case Zero => Nil
+                case Only(hs) => req.headers.filter(h => hs.contains(h.name))
+              }
+              Some(req.withUri(location.value).withHeaders(headers).withMethod(method))
+            } else {
+              None
+            }
+          })
+        }
+      } else {
+        None
+      }
+    }
+
+  def sameOrigin(u1: Uri, u2: Uri): Boolean = {
+    if (u2.isRelative) true
+    else if (u1.isAbsolute && u2.isAbsolute && u1.authority == u2.authority) true
+    else false
+  }
+
+  trait OpOutPrinter[F, _ <: Outlet[F]] {
+    def print(t: F): Unit
+  }
+
+  trait OpInPrinter[T <: Inlet[_]] {
+    def print(): Unit
+  }
```

Comment:
I'll see later if we really need those but in general let's not add ad-hoc in-middle-of-other-class type classes for printing :)